sdk/core/azure-core-amqp/src/models/amqp_header.cpp

```diff
@@ -0,0 +1,154 @@
+        && this->TimeToLive.HasValue() == that.TimeToLive.HasValue()
+        && (this->TimeToLive.HasValue() ? this->TimeToLive.Value() == that.TimeToLive.Value()
+                                        : true);
+  }
+
+  MessageHeader _internal::MessageHeaderFactory::FromUamqp(UniqueMessageHeaderHandle const& handle)
+  {
+    MessageHeader rv;
+    bool boolValue;
+    if (!header_get_durable(handle.get(), &boolValue))
+    {
+      rv.Durable = boolValue;
+    }
+
+    uint8_t uint8Value;
+    if (!header_get_priority(handle.get(), &uint8Value))
+    {
+      rv.Priority = uint8Value;
+    }
+
+    milliseconds millisecondsValue;
+    if (!header_get_ttl(handle.get(), &millisecondsValue))
+    {
+      rv.TimeToLive = std::chrono::milliseconds(millisecondsValue);
+    }
+
+    if (!header_get_first_acquirer(handle.get(), &boolValue))
+    {
+      rv.IsFirstAcquirer = boolValue;
+    }
+
+    uint32_t uint32Value;
+    if (!header_get_delivery_count(handle.get(), &uint32Value))
+    {
+      rv.DeliveryCount = uint32Value;
+    }
+    return rv;
+  }
+
+  UniqueMessageHeaderHandle _internal::MessageHeaderFactory::ToUamqp(MessageHeader const& header)
+  {
+    UniqueMessageHeaderHandle rv{header_create()};
+    if (header.Durable)
+    {
+      if (header_set_durable(rv.get(), header.Durable))
+      {
+        throw std::runtime_error("Could not set durable value.");
+      }
+    }
+    if (header.Priority != 4)
+    {
+      if (header_set_priority(rv.get(), header.Priority))
+      {
+        throw std::runtime_error("Could not set priority value.");
+      }
+    }
+    if (header.TimeToLive.HasValue())
+    {
+      if (header_set_ttl(rv.get(), static_cast<milliseconds>(header.TimeToLive.Value().count())))
+      {
+        throw std::runtime_error("Could not set header TTL."); // LCOV_EXCL_LINE
+      }
+    }
+
+    if (header.IsFirstAcquirer)
+    {
+      if (header_set_first_acquirer(rv.get(), header.IsFirstAcquirer))
+      {
+        throw std::runtime_error("Could not set first acquirer value.");
+      }
+    }
+    if (header.DeliveryCount != 0)
+    {
+      if (header_set_delivery_count(rv.get(), header.DeliveryCount))
+      {
+        throw std::runtime_error("Could not set delivery count value.");
+      }
+    }
+
+    return rv;
+  }
+
+  std::ostream& operator<<(std::ostream& os, MessageHeader const& header)
+  {
+    os << "Header{";
+    os << "durable=" << header.Durable;
+    os << ", priority=" << std::dec << static_cast<int>(header.Priority);
+    if (header.TimeToLive.HasValue())
+    {
+      os << ", ttl=" << header.TimeToLive.Value().count() << " milliseconds";
+    }
+    os << ", firstAcquirer=" << header.IsFirstAcquirer;
+    os << ", deliveryCount=" << header.DeliveryCount;
+    os << "}";
+    return os;
+  }
+
+  bool MessageHeader::ShouldSerialize() const noexcept
+  {
+    return Durable || (Priority != 4) || TimeToLive.HasValue() || IsFirstAcquirer
+        || (DeliveryCount != 0);
+  }
```

Comment:
What defines this logic for whether `MessageHeader` should serialize or not?

Reply:
If the header is marked as being Durable, if the priority is not 4, if the TTL has a value, if IsFirstAcquirer isn't set, or if DeliveryCount is not 0, then the message header must be serialized.

Each of these values is the default value for those properties as defined by the AMQP specification, if the value of the field departs from the default value, then it must be serialized, otherwise it should not be serialized.

Also, see #4503